File: C++/155 - Min Stack.cpp
/*
 155. Min Stack (E)

 Notes:
    Use two vector/Stack to address elements and minimum element
    T: O(n)
    S: O(n)
*/

// Vector Approach
class MinStack {
private:
    vector<int> stk;
    vector<int> minStk;
    
public:
    void push(int x) {
        stk.push_back(x);
        if (minStk.size())
            minStk.push_back(min(x, minStk.back()));
        else
            minStk.push_back(x);
    }
    
    void pop() {
        stk.pop_back();
        minStk.pop_back();
    }
    
    int top() {
        return stk.back();
    }
    
    int getMin() {
        return minStk.back();
    }
};

// Stack STL Approach
class MinStack {
private:
    stack<int> stk, minStk;

public:
    /** initialize your data structure here. */
    MinStack() {

    }

    void push(int x) {
        stk.push(x);

        if (minStk.empty() || x <= minStk.top())
            minStk.push(x);
    }

    void pop() {
        int top = stk.top();
        stk.pop();

        if (top == minStk.top())
            minStk.pop();
    }

    int top() {
        return stk.top();
    }

    int getMin() {
        return minStk.top();
    }
};

File: JAVA/155 - Min Stack.java
/*
 155. Min Stack (E)

 Approach 1 : Stack
 Use two stacks, one as the normal stack, one as the min nodes stack
 Time: O(n)
 Space: O(2n)

 Approach 2 : Custom wrap class
 Use a class to wrap the node val and min val
 Time: O(n)
 Space: O(n)

*/

/* Approach 1 */
class MinStack {
    Stack<Integer> stack;
    Stack<Integer> minStack;

    /** initialize your data structure here. */
    public MinStack() {
        stack = new Stack<Integer>();
        minStack = new Stack<Integer>();
    }
    
    public void push(int x) {
        stack.push(x);
        if (minStack.isEmpty()) 
            minStack.push(x);
        else 
            minStack.push(Math.min(x, minStack.peek()));
    }
    
    public void pop() {
        stack.pop();
        minStack.pop();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return minStack.peek();
    }
}


/* Approach 2 */
class MinStack {
    // Custom class
    class Node {
        int val, min;
        public Node(int val, int min) {
            this.val = val;
            this.min = min;
        }
    }
    
    Stack<Node> stack = new Stack<Node>();

    public MinStack() {}

    public void push(int x) {
        if (stack.isEmpty()) 
            stack.push(new Node(x, x));
        else
            stack.push(new Node(x, Math.min(x, stack.peek().min)));
    }
    
    public void pop() {
        stack.pop();
    }
    
    public int top() {
        return stack.peek().val;
    }
    
    public int getMin() {
        return stack.peek().min;
    }
}